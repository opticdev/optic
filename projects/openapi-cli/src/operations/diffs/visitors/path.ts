import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
