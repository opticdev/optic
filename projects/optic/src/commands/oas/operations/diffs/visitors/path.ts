import { OpenAPIV3 } from '../../../specs';
import {
  OperationDiffResult,
  OperationDiffResultKind,
} from '../../../../capture/patches/patchers/spec/types';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
