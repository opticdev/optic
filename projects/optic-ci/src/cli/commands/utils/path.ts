import path from 'path';
import { promisify } from 'util';
import { exec as callbackExec } from 'child_process';
const exec = promisify(callbackExec);

export const getRelativeRepoPath = (
  relativePath: string,
  gitRootPath: string | false
): string => {
  const absolutePath = path.join(process.cwd(), relativePath);

  if (!gitRootPath) {
    return absolutePath;
  } else {
    return absolutePath.startsWith(gitRootPath)
      ? absolutePath.replace(gitRootPath, '')
      : absolutePath;
  }
};

export const getGitRootPath = async () => {
  const { stdout: gitRootUntrimmed } = await exec(
    'git rev-parse --show-toplevel'
  );
  const gitRoot = gitRootUntrimmed.trim();
  return gitRoot;
};
