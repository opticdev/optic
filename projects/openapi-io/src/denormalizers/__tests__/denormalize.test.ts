import { test, expect, describe } from '@jest/globals';
import path from 'path';

import { denormalize } from '../denormalize';
import {
  ParseOpenAPIResult,
  parseOpenAPIWithSourcemap,
} from '../../parser/openapi-sourcemap-parser';

function prepSnapshot(result: ParseOpenAPIResult) {
  const cwd = process.cwd();
  result.sourcemap.files.forEach((i) => {
    i.path = i.path.split(cwd)[1];
    // @ts-ignore
    i.index = null;
  });

  result.sourcemap.rootFilePath = result.sourcemap.rootFilePath.split(cwd)[1];

  result.sourcemap.files = result.sourcemap.files.sort((a, b) =>
    a.path.toLocaleLowerCase().localeCompare(b.path.toLocaleLowerCase())
  );
  return result;
}
describe('denormalize', () => {
  test('denormalizes shared path parameters', async () => {
    const specPath = path.resolve(
      'src/denormalizers/__tests__/specs/openapi.yaml'
    );
    const spec = await parseOpenAPIWithSourcemap(specPath);

    const denormalized = denormalize(spec);

    expect(prepSnapshot(denormalized)).toMatchSnapshot();
  });

  describe('allOf merging', () => {
    test.each([
      [
        'merges allOf when all items are objects',
        'src/denormalizers/__tests__/specs/allOf/single-allof.yaml',
      ],
      [
        'does not merge allOf when all items are not all objects',
        'src/denormalizers/__tests__/specs/allOf/no-merge.yaml',
      ],
      [
        'merges nested allOf',
        'src/denormalizers/__tests__/specs/allOf/nested.yaml',
      ],
      [
        'merges allOfs in type array object / items',
        'src/denormalizers/__tests__/specs/allOf/in-type-array.yaml',
      ],
      [
        'merges allOf with only one item',
        'src/denormalizers/__tests__/specs/allOf/single-child.yaml',
      ],
    ])('%s', async (_, openapiFilePath) => {
      const spec = await parseOpenAPIWithSourcemap(
        path.resolve(openapiFilePath)
      );
      const warnings: string[] = [];
      const denormalized = denormalize(spec, warnings);

      expect(prepSnapshot(denormalized)).toMatchSnapshot();
      expect(warnings).toMatchSnapshot('warnings');
    });
  });
});
