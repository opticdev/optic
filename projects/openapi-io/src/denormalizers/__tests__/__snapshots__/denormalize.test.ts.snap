// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`denormalize v2 denormalizes shared path parameters 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "paths": {
      "/example": {
        "get": {
          "parameters": [
            {
              "description": "something here",
              "in": "query",
              "name": "something-else",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "should-not-override",
              "in": "query",
              "name": "should-not-override",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
          ],
          "produces": [
            "application/json",
            "application/json+something",
          ],
          "responses": {
            "200": {
              "description": "some thing",
            },
          },
        },
        "post": {
          "parameters": [
            {
              "description": "should not be overriden",
              "in": "query",
              "name": "should-not-override",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something-else",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
          ],
          "produces": [
            "application/json+something",
          ],
          "responses": {
            "200": {
              "description": "some thing",
            },
          },
        },
      },
    },
    "swagger": "2.0.1",
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "swagger: '2.0.1'
produces: ['application/json+something']
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    parameters:
      - name: something-else
        in: query
        required: true
        description: something here
        schema:
          type: string
          format: uuid
          example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      - name: should-not-override
        in: query
        required: true
        description: should-not-override
        schema:
          type: string
          format: uuid
          example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      - $ref: parameters.yml#/something
    get:
      produces: ['application/json']
      responses:
        200:
          description: 'some thing'
    post:
      parameters:
        - name: should-not-override
          in: query
          required: true
          description: should not be overriden
          schema:
            type: string
            format: uuid
            example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      responses:
        200:
          description: 'some thing'
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v2/openapi.yaml",
        "sha256": "8d13ebe93e4152fe53308e1d7365040aeb4828fb99b79c022e4afb950a9477d5",
      },
      {
        "contents": "something:
  name: something
  in: query
  required: true
  description: something here
  schema:
    type: string
    format: uuid
    example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v2/parameters.yml",
        "sha256": "85d23de44c126a054df04ca9720a69b0842935868c6d0a771baa37f0bee832ba",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/parameters/0": [
        0,
        "/paths/~1example/parameters/0",
      ],
      "/paths/~1example/get/parameters/1": [
        0,
        "/paths/~1example/parameters/1",
      ],
      "/paths/~1example/get/parameters/2": [
        1,
        "/paths/~1example/parameters/2",
      ],
      "/paths/~1example/parameters/2": [
        1,
        "/something",
      ],
      "/paths/~1example/post/parameters/1": [
        0,
        "/paths/~1example/parameters/0",
      ],
      "/paths/~1example/post/parameters/2": [
        1,
        "/paths/~1example/parameters/2",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v2/openapi.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging does not merge allOf when all items are not all objects 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/example": {
        "get": {
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "a": {
                      "type": "string",
                    },
                    "b": {
                      "type": "number",
                    },
                    "c": {
                      "type": "boolean",
                    },
                    "d": {
                      "properties": {
                        "a": {
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                  },
                  "type": "object",
                },
              },
            },
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "type": "string",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.0.1'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    get:
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                a:
                  type: string
                b:
                  type: number
                c:
                  type: boolean
                d:
                  allOf:
                    - type: object
                      properties:
                        a:
                          type: string
                    - type: string
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                type: 'string'
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/allOf/no-merge.yaml",
        "sha256": "4921acb94767de410927bd6751ac8d9f712c11dc70dcb260236a6684df8dcdb2",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/requestBody/content/application~1json/schema/properties/d/properties/a": [
        0,
        "/paths/~1example/get/requestBody/content/application~1json/schema/properties/d/allOf/0/properties/a",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/allOf/no-merge.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging does not merge allOf when all items are not all objects: warnings 1`] = `
[
  "invalid allOf variant at [4msrc/denormalizers/__tests__/specs/v3/allOf/no-merge.yaml:26:578[24m (/paths/~1example/get/requestBody/content/application~1json/schema/properties/d/allOf/1)",
]
`;

exports[`denormalize v3 allOf merging merges allOf when all items are objects 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/example": {
        "get": {
          "parameters": [],
          "requestBody": null,
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "properties": {
                      "a": {
                        "properties": {
                          "e": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "b": {
                        "type": "number",
                      },
                      "c": {
                        "type": "string",
                      },
                      "d": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "a",
                      "c",
                    ],
                    "type": "object",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.0.1'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    get:
      requestBody:
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                allOf:
                  - type: object
                    required:
                      - a
                    properties:
                      a:
                        type: object
                        properties:
                          e:
                            type: string
                  - type: object
                    required:
                      - c
                    properties:
                      a:
                        type: number
                        description: 'should not show up because this is after a previous key'
                      b:
                        type: number
                      c:
                        type: string
                  - type: object
                    properties:
                      d:
                        type: string
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/allOf/single-allof.yaml",
        "sha256": "16eba8b2fa4bae164c1443f0bdccaa8cdf1482e2a0a805bb215c0e219348455a",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/c": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/1/properties/c",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/d": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/2/properties/d",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/required/0": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/0/required/0",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/required/1": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/1/required/0",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/allOf/single-allof.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging merges allOf when all items are objects: warnings 1`] = `[]`;

exports[`denormalize v3 allOf merging merges allOf with only one item 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/example": {
        "get": {
          "parameters": [],
          "requestBody": null,
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "format": "uuid",
                    "type": "string",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.0.1'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    get:
      requestBody:
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                allOf:
                  - type: string
                    format: uuid
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/allOf/single-child.yaml",
        "sha256": "6536b93a83f2b2a5d776abac8830ba6750e4ba08056b78096b9c7ebf5653849c",
      },
    ],
    "refMappings": {},
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/allOf/single-child.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging merges allOf with only one item: warnings 1`] = `[]`;

exports[`denormalize v3 allOf merging merges allOfs in type array object / items 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.1.0",
    "paths": {
      "/example": {
        "get": {
          "parameters": [],
          "requestBody": null,
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "items": {
                      "properties": {
                        "a": {
                          "type": "string",
                        },
                        "b": {
                          "type": "number",
                        },
                        "c": {
                          "type": "string",
                        },
                        "d": {
                          "properties": {
                            "a": {
                              "type": "string",
                            },
                            "b": {
                              "type": "string",
                            },
                          },
                          "type": "object",
                        },
                        "e": {
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "properties": {
                      "inside_object": {
                        "properties": {
                          "a": {
                            "type": "string",
                          },
                          "b": {
                            "type": "number",
                          },
                          "c": {
                            "type": "string",
                          },
                          "d": {
                            "properties": {
                              "a": {
                                "type": "string",
                              },
                              "b": {
                                "type": "string",
                              },
                            },
                            "type": "object",
                          },
                          "e": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                    },
                    "type": [
                      "object",
                      "array",
                    ],
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.1.0'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    get:
      requestBody:
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                type: ['object', 'array']
                items:
                  allOf:
                    - type: object
                      properties:
                        a:
                          type: string
                    - type: object
                      properties:
                        b:
                          type: number
                        c:
                          type: string
                    - type: object
                      properties:
                        d:
                          allOf:
                            - type: 'object'
                              properties:
                                a:
                                  type: string
                            - type: 'object'
                              properties:
                                b:
                                  type: string
                        e:
                          type: string
                properties:
                  inside_object:
                    allOf:
                      - type: object
                        properties:
                          a:
                            type: string
                      - type: object
                        properties:
                          b:
                            type: number
                          c:
                            type: string
                      - type: object
                        properties:
                          d:
                            allOf:
                              - type: 'object'
                                properties:
                                  a:
                                    type: string
                              - type: 'object'
                                properties:
                                  b:
                                    type: string
                          e:
                            type: string
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/allOf/in-type-array.yaml",
        "sha256": "06b1b6e709c37580805955ed027bd8c659bf85e9a8b2333c7d6e21475f10687e",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/c": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/1/properties/c",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/d": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/2/properties/d",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/d/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/2/properties/d/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/d/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/2/properties/d/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/items/properties/e": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/items/allOf/2/properties/e",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/c": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/1/properties/c",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/d": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/2/properties/d",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/d/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/2/properties/d/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/d/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/2/properties/d/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/properties/e": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/properties/inside_object/allOf/2/properties/e",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/allOf/in-type-array.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging merges allOfs in type array object / items: warnings 1`] = `[]`;

exports[`denormalize v3 allOf merging merges nested allOf 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/example": {
        "get": {
          "parameters": [],
          "requestBody": null,
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "properties": {
                      "a": {
                        "type": "string",
                      },
                      "b": {
                        "type": "number",
                      },
                      "c": {
                        "type": "string",
                      },
                      "d": {
                        "properties": {
                          "a": {
                            "type": "string",
                          },
                          "b": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "e": {
                        "type": "string",
                      },
                      "x": {
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.0.1'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    get:
      requestBody:
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                type: object
                allOf:
                  - allOf:
                      - type: object
                        properties:
                          x:
                            type: string
                  - type: object
                    properties:
                      a:
                        type: string
                  - type: object
                    properties:
                      b:
                        type: number
                      c:
                        type: string
                  - type: object
                    properties:
                      d:
                        allOf:
                          - type: 'object'
                            properties:
                              a:
                                type: string
                          - type: 'object'
                            properties:
                              b:
                                type: string
                      e:
                        type: string
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/allOf/nested.yaml",
        "sha256": "f40a14b89b16b1dbccde926823ead4a870178e400878f5114b21936725d99527",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/0/properties/x": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/0/allOf/0/properties/x",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/1/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/2/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/c": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/2/properties/c",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/d": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/3/properties/d",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/d/properties/a": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/3/properties/d/allOf/0/properties/a",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/d/properties/b": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/3/properties/d/allOf/1/properties/b",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/e": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/3/properties/e",
      ],
      "/paths/~1example/get/responses/200/content/application~1json/schema/properties/x": [
        0,
        "/paths/~1example/get/responses/200/content/application~1json/schema/allOf/0/properties/x",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/allOf/nested.yaml",
  },
}
`;

exports[`denormalize v3 allOf merging merges nested allOf: warnings 1`] = `[]`;

exports[`denormalize v3 denormalizes shared path parameters 1`] = `
{
  "jsonLike": {
    "info": {
      "title": "some thing",
      "version": "v0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/example": {
        "get": {
          "parameters": [
            {
              "description": "something here",
              "in": "query",
              "name": "something-else",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "should-not-override",
              "in": "query",
              "name": "should-not-override",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "type": "string",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
        "post": {
          "parameters": [
            {
              "description": "should not be overriden",
              "in": "query",
              "name": "should-not-override",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something-else",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
            {
              "description": "something here",
              "in": "query",
              "name": "something",
              "required": true,
              "schema": {
                "example": "d5b640e5-d88c-4c17-9bf0-93597b7a1ce2",
                "format": "uuid",
                "type": "string",
              },
            },
          ],
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "schema": {
                    "type": "string",
                  },
                },
              },
              "description": "some thing",
              "headers": {},
            },
          },
        },
      },
    },
  },
  "sourcemap": JsonSchemaSourcemap {
    "files": [
      {
        "contents": "openapi: '3.0.1'
info:
  title: 'some thing'
  version: 'v0'
paths:
  /example:
    parameters:
      - name: something-else
        in: query
        required: true
        description: something here
        schema:
          type: string
          format: uuid
          example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      - name: should-not-override
        in: query
        required: true
        description: should-not-override
        schema:
          type: string
          format: uuid
          example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      - $ref: parameters.yml#/something
    get:
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                type: 'string'
    post:
      parameters:
        - name: should-not-override
          in: query
          required: true
          description: should not be overriden
          schema:
            type: string
            format: uuid
            example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
      responses:
        200:
          description: 'some thing'
          content:
            application/json:
              schema:
                type: 'string'
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/openapi.yaml",
        "sha256": "44bb048800379bdd9651daa9c6a380dbd68232443881cbbaa0c155240e45c001",
      },
      {
        "contents": "something:
  name: something
  in: query
  required: true
  description: something here
  schema:
    type: string
    format: uuid
    example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
",
        "index": null,
        "path": "/src/denormalizers/__tests__/specs/v3/parameters.yml",
        "sha256": "85d23de44c126a054df04ca9720a69b0842935868c6d0a771baa37f0bee832ba",
      },
    ],
    "refMappings": {
      "/paths/~1example/get/parameters/0": [
        0,
        "/paths/~1example/parameters/0",
      ],
      "/paths/~1example/get/parameters/1": [
        0,
        "/paths/~1example/parameters/1",
      ],
      "/paths/~1example/get/parameters/2": [
        1,
        "/paths/~1example/parameters/2",
      ],
      "/paths/~1example/parameters/2": [
        1,
        "/something",
      ],
      "/paths/~1example/post/parameters/1": [
        0,
        "/paths/~1example/parameters/0",
      ],
      "/paths/~1example/post/parameters/2": [
        1,
        "/paths/~1example/parameters/2",
      ],
    },
    "rootFilePath": "/src/denormalizers/__tests__/specs/v3/openapi.yaml",
  },
}
`;
